### PR TITLE
Warning for missing .cmx files

### PR DESCRIPTION
**Pull request by lpw25**
This patch adds a warning for missing cmx files during cross-module inlining. This is useful because currently a misspecified build can silently damage the performance of programs.

The patch also extends the `-opaque` option to work on .mli files. Using `-opaque` on an .ml file means that its implementation should not be considered for cross-module inlining. Using `-opaque` on an .mli file means that no implementation of that interface should be considered for cross-module inlining.

If an interface is marked as `opaque` then the compiler does not search for a .cmx file, and so the new warning will not be triggered. This means that modules which are intended to be chosen at link-time (and so deliberately hide .cmx files) can avoid unnecessarily triggering the warning.

As an illustration of the utility of this new warning, not all cmx files were installed by the OCaml distribution, and the patch also adjusts the installation rules to install these files (which were in compiler-libs and ocamldoc).

### CODE REVIEW

**Comment by gasche** [comment]
> Using -opaque on an .ml file means that its implementation should not be considered for cross-module inlining. Using -opaque on an .mli file means that no implementation of that interface should be considered for cross-module inlining.

I have a more semantic/indirect definition for what `-opaque` is, or at least should be: `-opaque` should guarantee that, if a module that is compiled with `-opaque` changes its implementation and gets recompiled, then its dependencies do not need to be recompiled as well. For now I think that both definitions coincide, but that is the "no implementation-dependency on opaque modules" that needs to be preserved in the future.

File: typing/cmt_format.ml [hunk 1]
```diff
@@ -164,6 +164,12 @@ let record_value_dependency vd1 vd2 =
 let save_cmt filename modname binary_annots sourcefile initial_env sg =
   if !Clflags.binary_annotations && not !Clflags.print_types then begin
     let imports = Env.imports () in
+    let flags =
+      if !Clflags.recursive_types then [Cmi_format.Rectypes] else []
+    in
+    let flags =
+      if !Clflags.opaque then Cmi_format.Opaque :: flags else flags
+    in
```
Comment (gasche) [inline typing/cmt_format.ml:172]: This code is duplicated twice, here and in Env. Would it make sense to factorize it into a `Cmi_format.global_pers_flags ()` function?

Also with my nitpicking hat on, I think you could factorize this as

``` ocaml
  []
  @ (if !Clflags.recursive_types then [Cmi_format.Rectypes] else [])
  @ (if !Clflags.opaque then [Cmi_format.Opaque] else [])
```

File: typing/env.ml [hunk 2]
```diff
@@ -315,9 +315,15 @@ let imported_units = ref StringSet.empty
 let add_import s =
   imported_units := StringSet.add s !imported_units
 
+let opaque_units = ref StringSet.empty
```
Comment (gasche) [inline typing/env.ml]: `opaque_imported_units` would have made the code easier to read for me.

File: typing/env.mli [hunk 3]
```diff
@@ -170,6 +170,9 @@ val crc_of_unit: string -> Digest.t
 
 val imports: unit -> (string * Digest.t option) list
 
+(* [is_opaque md] returns true if [md] is an opaque imported module  *)
+val is_opaque: string -> bool
```
Comment (gasche) [inline typing/env.mli]: I think `opaque_import` or `is_opaque_import` would be a clearer name. It is not clear from the type _what_ is or is not opaque.

File: utils/warnings.ml [hunk 4]
```diff
@@ -426,6 +428,8 @@ let message = function
       Printf.sprintf "the %S attribute is used more than once on this expression" attr_name
   | Inlining_impossible reason ->
       Printf.sprintf "Inlining impossible in this context: %s" reason
+  | No_cmx_file name ->
+      Printf.sprintf "no cmx file was found in path for module " ^ name
```
Comment (gasche) [inline utils/warnings.ml]: Would there be a way to mention in the message that `-opaque` could be used, that would not make it the Daughter of All Warning Messages?

**Comment by gasche** [comment]
I like the feature, and the code seems fairly simple and very reasonable. In favor.

File: utils/warnings.ml (hunk 4 above)
Comment (gasche) [inline utils/warnings.ml]:
```
    Printf.sprintf
      "no cmx file was found in path for module %s, \
       yet it was not compiled with -opaque" name
```

?

Comment (gasche) [inline utils/warnings.ml]: (Looking back at your line of code, there is something troubling about using both `sprintf` and `(^)`.)

Comment (Drup) [inline utils/warnings.ml]: And with a constant format.

Comment (lpw25) [inline utils/warnings.ml]: I would suggest:

``` ocaml
Printf.sprintf
      "no cmx file was found in path for module %s, \
       and it's interface was not compiled with -opaque" name
```

Comment (gasche) [inline utils/warnings.ml]: Excellent, please do!

**Comment by lpw25** [comment]
> I have a more semantic/indirect definition for what -opaque is, or at least should be: -opaque should guarantee that, if a module that is compiled with -opaque changes its implementation and gets recompiled, then its dependencies do not need to be recompiled as well.

I would actually go further and define it in terms of being able to choose between different implementations at link-time.

**Comment by lpw25** [comment]
I'll tidy up the code, warning message and changes file this evening.

**Comment by gasche** [comment]
I wanted to merge this PR right now, but in fact it now conflicts with @alainfrisch 's change on compilation unit deprecation. Could you rebase the patchset against the current trunk? In addition, it would be nice if you could:
- squash the minor fixup commits (it's nice to have separate commits for modified install rules, but the rest is just noise)
- add a very small testsuite exercising both `-opaque` on `ml` files (the test could simply modify the source, rebuild and relink as a way to check that a dependency was really seing the compilation unit as opaque), on `mli` files (same thing, but with `-opaque` passed earlier in the compilation chain), and the warning.

File: utils/warnings.ml [hunk 5]
```diff
@@ -426,6 +428,10 @@ let message = function
       Printf.sprintf "the %S attribute is used more than once on this expression" attr_name
   | Inlining_impossible reason ->
       Printf.sprintf "Inlining impossible in this context: %s" reason
+  | No_cmx_file name ->
+      Printf.sprintf
+        "no cmx file was found in path for module %s, \
+         and it's interface was not compiled with -opaque" name
```
Comment (damiendoligez) [inline utils/warnings.ml]: Should be "its".

**Comment by lpw25** [comment]
I've fixed the message, rebased and added tests for both `-opaque` and the new warning.

File: typing/env.ml [hunk 6]
```diff
@@ -1778,10 +1794,15 @@ let save_signature_with_imports ~deprecated sg modname filename imports =
   Btype.cleanup_abbrev ();
   Subst.reset_for_saving ();
   let sg = Subst.signature (Subst.for_saving Subst.identity) sg in
+  let flags =
+    List.concat [
+      if !Clflags.recursive_types then [Cmi_format.Rectypes] else [];
+      if !Clflags.opaque then [Cmi_format.Opaque] else [];
+      match deprecated with Some s -> [Deprecated s] | None -> [];
```
Comment (gasche) [inline typing/env.ml]: this last line should be parenthesized or use `begin .. end`.

Comment (lpw25) [inline typing/env.ml]: Done

**Comment by gasche** [comment]
Nice PR, thanks!

**Comment by gasche** [comment]
@lpw25 with more usage experience, there is one aspect of this change that I'm not fully satisfied with: it is the fact that `-opaque` is only supported by `ocamlopt`, not `ocamlc`. This means that people that want to compile a `.mli` to produce an opaque `.cmi` have to use `ocamlopt -opaque`, and cannot use `ocamlc -opaque` for this purpose. This changes the previously-true property that either compiler can be used to compile interfaces, and I wouldn't be surprised if some makefiles or build systems relied on it.

I think that with your change, we should add `-opaque` as an option to ocamlc as well. If you agree, would you maybe have the time to do this change in the following days? (I probably don't, but we could consider asking someone else.)

**Comment by lpw25** [comment]
I completely agree, in fact I consider this to be a bug as it was not intentional. However, I'm not sure that I'll have time to look at this in the next few days.

**Comment by damiendoligez** [comment]
ping

**Pull request merged**